File: clients/android/app/src/main/kotlin/com/tdg/loader/Receipt.kt
package com.tdg.loader

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * What this app wrote, and where.
 *
 * Two jobs, both essential. It makes a fill **resumable** — Android will kill a
 * twenty-minute foreground service eventually, and restarting a 64 GB transfer
 * from zero is not acceptable. And it makes **wipe** exact: the app deletes the
 * content URIs it recorded and nothing else.
 *
 * Keyed on the MediaStore URI rather than the filename, deliberately. Filenames
 * are not a reliable handle for an imported asset — iOS renames every one — and
 * the URI is what ContentResolver.delete actually needs.
 */
class Receipt(context: Context, val jobId: String) {

    private val file = File(context.filesDir, "receipt-$jobId.json")
    private val entries = LinkedHashMap<String, String>()   // name -> content uri

    init { load() }

    val count: Int get() = entries.size

    fun has(name: String) = entries.containsKey(name)

    fun uris(): List<String> = entries.values.toList()

    @Synchronized
    fun add(name: String, uri: String) {
        entries[name] = uri
        save()
    }

    @Synchronized
    fun remove(name: String) {
        entries.remove(name)
        save()
    }

    @Synchronized
    fun clear() {
        entries.clear()
        file.delete()
    }

    private fun load() {
        if (!file.exists()) return
        try {
            val arr = JSONArray(file.readText())
            for (i in 0 until arr.length()) {
                val o = arr.getJSONObject(i)
                entries[o.getString("name")] = o.getString("uri")
            }
        } catch (e: Exception) {
            // A torn write from a killed process. Losing the receipt costs a
            // re-download; a crash loop on startup would cost far more.
            file.delete()
            entries.clear()
        }
    }

    private fun save() {
        val arr = JSONArray()
        entries.forEach { (name, uri) ->
            arr.put(JSONObject().put("name", name).put("uri", uri))
        }
        val tmp = File(file.parentFile, file.name + ".tmp")
        tmp.writeText(arr.toString())
        tmp.renameTo(file)      // atomic: a receipt is never half-written
    }

    companion object {
        fun knownJobs(context: Context): List<String> =
            context.filesDir.listFiles()
                ?.mapNotNull { f ->
                    f.name.takeIf { it.startsWith("receipt-") && it.endsWith(".json") }
                        ?.removePrefix("receipt-")?.removeSuffix(".json")
                }
                ?.sorted().orEmpty()
    }
}

File: clients/ios/TdgLoader/Receipt.swift
import Foundation

/// What this app added, and what Photos called it.
///
/// Keyed on the `PHAsset` **localIdentifier**, never the filename: Photos
/// renames every imported asset to `IMG_NNNN`, so a filename-keyed wipe would
/// find nothing. The identifier is also exactly what `deleteAssets` needs.
///
/// It makes a fill resumable after the app is killed, and it makes wipe exact —
/// the app deletes the identifiers it recorded and nothing else in the library
/// is touched.
final class Receipt {
    let jobId: String
    private let url: URL
    private var byName: [String: String] = [:]      // file name -> localIdentifier
    private let queue = DispatchQueue(label: "tdg.receipt")

    init(jobId: String) {
        self.jobId = jobId
        let dir = FileManager.default.urls(for: .applicationSupportDirectory,
                                           in: .userDomainMask)[0]
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        self.url = dir.appendingPathComponent("receipt-\(jobId).json")
        load()
    }

    var count: Int { queue.sync { byName.count } }
    var identifiers: [String] { queue.sync { Array(byName.values) } }

    func has(_ name: String) -> Bool { queue.sync { byName[name] != nil } }

    func add(_ pairs: [(String, String)]) {
        queue.sync {
            for (name, id) in pairs { byName[name] = id }
            save()
        }
    }

    func clear() {
        queue.sync {
            byName.removeAll()
            try? FileManager.default.removeItem(at: url)
        }
    }

    private func load() {
        guard let data = try? Data(contentsOf: url),
              let o = try? JSONSerialization.jsonObject(with: data) as? [String: String]
        else { return }
        byName = o
    }

    /// Written to a sibling then moved, so a crash mid-write cannot leave a
    /// half-parsed receipt that would strand assets in the library.
    private func save() {
        guard let data = try? JSONSerialization.data(withJSONObject: byName) else { return }
        let tmp = url.deletingLastPathComponent()
            .appendingPathComponent("receipt-\(jobId).tmp")
        do {
            try data.write(to: tmp, options: .atomic)
            _ = try FileManager.default.replaceItemAt(url, withItemAt: tmp)
        } catch {
            try? data.write(to: url, options: .atomic)
        }
    }

    static func knownJobs() -> [String] {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory,
                                           in: .userDomainMask)[0]
        let names = (try? FileManager.default.contentsOfDirectory(atPath: dir.path)) ?? []
        return names.compactMap { n in
            guard n.hasPrefix("receipt-"), n.hasSuffix(".json") else { return nil }
            return String(n.dropFirst("receipt-".count).dropLast(".json".count))
        }.sorted()
    }
}
